compiler/rustc_codegen_ssa/src/mir/operand.rs

```diff
@@ -204,14 +204,25 @@ impl<'a, 'tcx, V: CodegenObject> OperandRef<'tcx, V> {
         let alloc_align = alloc.inner().align;
         assert!(alloc_align >= layout.align.abi);
 
+        // Returns `None` when the value is partially undefined or any byte of it has provenance.
+        // Otherwise returns the value or (if the entire value is undef) returns an undef.
         let read_scalar = |start, size, s: abi::Scalar, ty| {
+            let range = alloc_range(start, size);
             match alloc.0.read_scalar(
                 bx,
-                alloc_range(start, size),
+                range,
                 /*read_provenance*/ matches!(s.primitive(), abi::Primitive::Pointer(_)),
             ) {
-                Ok(val) => bx.scalar_to_backend(val, s, ty),
-                Err(_) => bx.const_poison(ty),
+                Ok(val) => Some(bx.scalar_to_backend(val, s, ty)),
+                Err(_) => {
+                    if alloc.0.provenance().range_empty(range, &bx.tcx())
+                        && alloc.0.init_mask().is_range_initialized(range).unwrap_err() == range
```

Comment:
Doesn't this ICE if the range is partially initialized?

Reply:
This is a bit subtle

```
    /// Checks whether the `range` is entirely initialized.
    ///
    /// Returns `Ok(())` if it's initialized. Otherwise returns a range of byte
    /// indexes for the first contiguous span of the uninitialized access.
    pub fn is_range_initialized(&self, range: AllocRange) -> Result<(), AllocRange> {
```

Since the `read_scalar` failed, and no relocations were involved, it must be that some part of it (or all) is undef.

Reply:
I added comments and tests

Reply:
Ah, I misunderstood exactly what `is_range_initialized` does. You get `Err` if any byte is uninit. Not sure how I misunderstood that at first.